src/backend/storage/ipc/procarray.c

```diff
@@ -2247,24 +2252,26 @@ updateSharedLocalSnapshot(DtxContextInfo *dtxContextInfo,
 					snapshot->xcnt,
 					snapshot->curcid)));
 
+	sz = EstimateSnapshotSpace(snapshot);
 	LWLockAcquire(SharedLocalSnapshotSlot->slotLock, LW_EXCLUSIVE);
 
-	SharedLocalSnapshotSlot->snapshot.xmin = snapshot->xmin;
-	SharedLocalSnapshotSlot->snapshot.xmax = snapshot->xmax;
-	SharedLocalSnapshotSlot->snapshot.xcnt = snapshot->xcnt;
-
-	if (snapshot->xcnt > 0)
+	if (sz > SharedLocalSnapshotSlot->snapshot_sz)
 	{
-		Assert(snapshot->xip != NULL);
-
-		ereport((Debug_print_full_dtm ? LOG : DEBUG5),
-				(errmsg("updateSharedLocalSnapshot count of in-doubt ids %u",
-						SharedLocalSnapshotSlot->snapshot.xcnt)));
-
-		memcpy(SharedLocalSnapshotSlot->snapshot.xip, snapshot->xip, snapshot->xcnt * sizeof(TransactionId));
+		if (SharedLocalSnapshotSlot->snapshot_handle != DSM_HANDLE_INVALID)
+			dsm_unpin_segment(SharedLocalSnapshotSlot->snapshot_handle);
+
+		snapshot_sgm = dsm_create(sz, 0);
+		dsm_pin_segment(snapshot_sgm);
+		dsm_pin_mapping(snapshot_sgm);
+		SharedLocalSnapshotSlot->snapshot_handle = dsm_segment_handle(snapshot_sgm);
+		SharedLocalSnapshotSlot->snapshot_sz = sz;
 	}
-	
-	SharedLocalSnapshotSlot->snapshot.curcid = snapshot->curcid;
+	else
+		snapshot_sgm = dsm_attach(SharedLocalSnapshotSlot->snapshot_handle);
```

Comment:
Is it possible `snapshot_sgm == NULL` here?

Reply:
No. The `snapshot_handle` is set iff `snapshot_sz` is not 0.